blockprod/src/detail/job_manager.rs

```diff
@@ -0,0 +1,358 @@
+        }
+    }
+
+    pub async fn new_job(
+        &self,
+        block_id: Id<GenBlock>,
+    ) -> Result<(JobKey, UnboundedReceiver<()>), JobManagerError> {
+        let (result_sender, result_receiver) = oneshot::channel();
+        let (cancel_sender, cancel_receiver) = unbounded_channel::<()>();
+
+        if self.new_job_sender.send((block_id, cancel_sender, result_sender)).is_err() {
+            return Err(JobManagerError::FailedToSendNewJobEvent);
+        }
+
+        tokio::select! {
+            result = result_receiver => match result {
+                Ok(v) => match v {
+                    Ok(v) => Ok((v, cancel_receiver)),
+                    Err(e) => Err(e),
+                },
+                Err(_) => Err(JobManagerError::FailedToCreateJob)
+            }
+        }
+    }
+
+    #[allow(clippy::type_complexity)]
+    fn new_job_handler(
+        jobs: &mut BTreeMap<JobKey, JobHandle>,
+        event: Option<(
+            Id<GenBlock>,
+            UnboundedSender<()>,
+            oneshot::Sender<Result<JobKey, JobManagerError>>,
+        )>,
+    ) {
+        if let Some((current_tip_id, cancel_sender, result_sender)) = event {
+            let job_key = JobKey::new(current_tip_id);
+
+            if jobs.contains_key(&job_key) {
+                if let Err(e) = result_sender.send(Err(JobManagerError::JobAlreadyExists)) {
+                    log::info!("Error sending new job exists error: {e:?}");
+                }
+            } else {
+                jobs.insert(job_key.clone(), JobHandle { cancel_sender });
+
+                if let Err(e) = result_sender.send(Ok(job_key)) {
+                    log::info!("Error sending new job event: {e:?}");
+                }
+            }
+        }
+    }
+
+    pub async fn stop_job(&mut self, job_key: Option<JobKey>) -> Result<usize, JobManagerError> {
+        let (result_sender, result_receiver) = oneshot::channel();
+
+        if self.stop_job_sender.send((job_key, result_sender)).is_err() {
+            return Err(JobManagerError::FailedToSendStopJobEvent);
+        }
+
+        tokio::select! {
+            result = result_receiver => match result {
+                Ok(v) => Ok(v),
+                Err(_) => Err(JobManagerError::FailedToStopJobs),
+            }
+        }
+    }
+
+    fn stop_job_handler(
+        jobs: &mut BTreeMap<JobKey, JobHandle>,
+        event: Option<(Option<JobKey>, oneshot::Sender<usize>)>,
+    ) {
+        if let Some((job_key, result_sender)) = event {
+            let mut stop_jobs = Vec::new();
+
+            match job_key {
+                Some(job_key) => {
+                    if let Some(job_handle) = jobs.remove(&job_key) {
+                        stop_jobs.push((job_key, job_handle));
+                    }
+                }
+                None => {
+                    while let Some((job_key, job_handle)) = jobs.pop_first() {
+                        stop_jobs.push((job_key, job_handle));
+                    }
```

Comment:
You can `std::mem::take` the `jobs` object to take ownership and then `.into_iter().collect()`

Reply:
This actually worked. Cool :)